File: src/solar/solar-common/w3ts/handles/handle.ts
/** @noSelfInFile **/
import DataBase from "@/DataBase";

// const map: WeakMap<handle, any> = new WeakMap<handle, any>();

export class Handle<T extends handle> {

    public readonly handle: T;
    private static initHandle: handle | undefined;

    protected constructor(handle?: T) {
        this.handle = handle === undefined ? Handle.initHandle as T : handle;
        // map.set(this.handle, this);
    }

    public get handleId() {
        return GetHandleId(this.handle);
    }

    public static initFromHandle(): boolean {
        return Handle.initHandle !== undefined;
    }

    protected static getObject(handle: handle, handleType?: string): any {
        // const obj = map.get(handle);
        // if (obj !== undefined) {
        //     return obj;
        // }
        if (!IsHandle(handle)) {
            let tb = debug.traceback();
            log.error("Handle.getObject():你传的handle为空或0！")
            log.error(tb)
            return null;
        }

        Handle.initHandle = handle;
        let handleObj = null
        if (handleType != null) {
            /** 从solarData里获取 包装对象 以保证每次获取的都是相同的对象*/
            let solarData = DataBase.getDataByHandle(handleType, handle, true);
            handleObj = solarData._sl_handleObj;
            //判断一下handle 是否相等 以免作者手动复制了solarData 把不属于自己缓存的包装对象复制过来了
            if (handleObj == null || handleObj.handle != handle) {
                handleObj = new this();
                solarData._sl_handleObj = handleObj;
            }
        } else {
            handleObj = new this();
        }
        Handle.initHandle = undefined;
        return handleObj;
    }

}

File: scripts/solar/solar-common/w3ts/handles/handle.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 2,["9"] = 2,["10"] = 6,["11"] = 6,["12"] = 6,["13"] = 11,["14"] = 12,["15"] = 11,["16"] = 20,["17"] = 21,["18"] = 20,["19"] = 24,["20"] = 29,["21"] = 30,["22"] = 31,["23"] = 32,["24"] = 33,["26"] = 36,["27"] = 37,["28"] = 38,["30"] = 40,["31"] = 41,["32"] = 43,["33"] = 44,["34"] = 45,["37"] = 48,["39"] = 50,["40"] = 51,["41"] = 24,["46"] = 17});
local ____exports = {}
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
____exports.Handle = __TS__Class()
local Handle = ____exports.Handle
Handle.name = "Handle"
function Handle.prototype.____constructor(self, handle)
    self.handle = handle == nil and ____exports.Handle.initHandle or handle
end
function Handle.initFromHandle(self)
    return ____exports.Handle.initHandle ~= nil
end
function Handle.getObject(self, handle, handleType)
    if not IsHandle(handle) then
        local tb = debug.traceback()
        log.error("Handle.getObject():你传的handle为空或0！")
        log.error(tb)
        return nil
    end
    ____exports.Handle.initHandle = handle
    local handleObj = nil
    if handleType ~= nil then
        --- 从solarData里获取 包装对象 以保证每次获取的都是相同的对象
        local solarData = DataBase:getDataByHandle(handleType, handle, true)
        handleObj = solarData._sl_handleObj
        if handleObj == nil or handleObj.handle ~= handle then
            handleObj = __TS__New(self)
            solarData._sl_handleObj = handleObj
        end
    else
        handleObj = __TS__New(self)
    end
    ____exports.Handle.initHandle = nil
    return handleObj
end
__TS__SetDescriptor(
    Handle.prototype,
    "handleId",
    {get = function(self)
        return GetHandleId(self.handle)
    end},
    true
)
return ____exports
